keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,359 @@
+    Kubelet makes sure that the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
+    Kubelet passes it as an mount option to all CSI calls for given volume.
+
+    After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
+    * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+    * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly or the volume was already mounted:
+        * If the volume supports SELinux (i.e. has `selabel` mount option in `/proc/mounts`), it passes ":Z" to the container runtime for the volume.
+            This is current kubelet behavior.
+        * If the volume does not support SELinux, it does not pass any ":Z" option to the container runtime.
+
+* Nothing changes if kubelet does not know the SELinux context of a pod (`SELinuxOptions` are empty),  kubelet passes ":Z" to the container runtime as today and lets the container runtime to choose a random label + relabel the volumes. 
+* Nothing changes if pod's `SELinuxRelabelPolicy` is `Always`, kubelet passes ":Z" to the container runtime as today.
+* Kubernetes validation check that `SELinuxRelabelPolicy` field can be used in a pod only when `SELinuxOptions` is set and SELinux label is known.
+
+#### Open Questions
+    When ... Pod/Container contains `SELinuxOptions`, ... [kubeket] tries to mount all volumes ... using `mount -o context=XYZ`.
+
+To mount with `-o context`, kubelet must know *all* `SELinuxOptions`, i.e. `User`, `Role`, `Type`, `Level`.
+Currently, when one or more of these fields are empty, the container runtime uses a default value set by OS.
+We need a way for kubelet to get these defaults.
+
+Similar problem arises with "*If the volume root has expected label*" above - kubelet must know complete SELinux label to be able to check it.
+
+### Shared volumes
+
+If a single volume that supports SELinux labels is being shared by multiple pods, each of them must have the same SELinux context.
+Currently, a running pod with context `A` will loose access to all files on a volume if a pod with context `B` starts and uses the same volume, because the container runtime relabels the volume for pod `B`.  
+This behavior changes with this KEP: kubelet mounts the volume with `-o context=A` for  the first pod.
+It tries to do the same for the second pod with `-o context=B`, however, the volume has already been mounted and kernel won't change the label of the volume. So pod `A` can access the volume, while pod `B` cannot.
+
+We don't think that this is a bug in the design.
```

Comment:
If the volume is labeled with an MCS Label of S0, then all containers can use it.
system_u:object_r:container_file_t:s0

Reply:
Great suggestion, added link to `go-selinux` and explained how the various label components are defaulted. In short, Level is never defaulted and must be non-empty in a Pod to trigger the new behavior. All other components can be empty.

Reply:
Fixed